### ISSUE TITLE
ElasticSearch auto-complete repository implementation

### ISSUE THREAD

**Issue by quantranhong1999** [linagora/tmail-backend#365]
## Why 
This work has been started partially by Giang. but it is not finished yet. We need to continue the work and finish it.

## How
- Open a PR on james-project (IndexCreationFactory) adding some static param fields for ngram tokenizer. cf https://github.com/linagora/tmail-backend/pull/329/commits/1f8cdfe224bd7c849b701f62e5632eeed6d622b0#r787560681
- Add the tokenizer parameter in Tmail contact document mapping: EmailAddressContactMappingFactory. Remind: switch other params to non hard code (use Static fields).
- Refine the existing ES implementation code and the contract tests.

## DoD
- Pass contract tests.

**Comment by chibenwa** [comment]
> Open a PR on james-project (IndexCreationFactory) adding some static param fields for ngram tokenizer. cf https://github.com/linagora/tmail-backend/commit/1f8cdfe224bd7c849b701f62e5632eeed6d622b0#r787560681

This might not be needed if we can use search_as_you_type` fields. Can you have a quick POC for the support of this on our various ElasticSearch versions?

**Comment by quantranhong1999** [comment]
`In the Elastic Stack 7.2 release, a new feature has been added to deal with typeahead in a really easy way, a data type called search_as_you_type. ` cf https://www.linkedin.com/pulse/new-elasticsearch-datatype-searchasyoutype-jess%C3%A9-peixoto/

I tested quickly using docker image;
- ES 6.3.2 does not support search_as_you_type
- ES 7.10.2 supports it (so should the same for OpenDistro).
  The thing here is with input: `hqtran@linagora.com` and using default query [at ES docs](https://www.elastic.co/guide/en/elasticsearch/reference/current/search-as-you-type.html). The queries start with the first character `h` like `h`, `hq`, `hqtran` is OK. But the queries with some thing like %X%: `qtr`, `tran`... do not return any hits.
  It looks like this search_as_you_type uses [Edge n-gram tokenizer](https://www.elastic.co/guide/en/elasticsearch/reference/current/analysis-edgengram-tokenizer.html) by default. We will need to specify another suitable tokenizer attached to search_as_you_type.

**Comment by quantranhong1999** [comment]
I think we should support the contact name alongside the contact address?
Contact 1:
- Name: Quan Tran
- Email address: hqtran@linagora.com

So when we type `quan` => it does suggest hqtran@linagora.com.

**Comment by chibenwa** [comment]
Ok thanks for checking.

I'm thus ok to refactor James to allow specifying additional tokenizers/analyzers upon mapping creation.

Please ensure the relevant ticket is up to date and added to the grooming.

**Comment by chibenwa** [comment]
> I think we should support the contact name alongside the contact address?

I like the idea.

Yes, when we have it, we can include firstname / surname.

 -> We can get it from the calendar app.
 -> We can get it from the LDAP though it likely means that we will need a new LSC script to account for this and that we can't instrument things at the level of the UsersRepository. 
 
We would then need a separated webadmin API to:

 - Create a contact
 - list contacts
 - Delete Contact
 - Update contact
 
+ a LSC script to do that.

 -> In the mailetcontainer for auto-collected contacts we "MIGHT" have such details at hand too is specified by the MUA.

**Comment by chibenwa** [comment]
We could have 1 index with two documents in it:

 - with `owner` field
 - with `domain` field

Upon query

`autocomplete(owner) OR autocomplete(domain)`

**Comment by quantranhong1999** [comment]
I did a quick LAB to try some index/query. Here is my current proposal:

## User private contact ES auto-complete
### Index
```json
{
	"settings": {
		"index": {
			"max_ngram_diff": 2
		},
		"analysis": {
			"analyzer": {
				"ngram_filter_analyzer": {
					"tokenizer": "standard",
					"filter": ["ngram_filter"]
				},
				"edge_ngram_filter_analyzer": {
					"tokenizer": "standard",
					"filter": ["edge_ngram_filter", "lowercase"]
				}
			},
			"filter": {
				"ngram_filter": {
					"type": "ngram",
					"min_gram": 3,
					"max_gram": 5
				},
				"edge_ngram_filter": {
					"type": "edge_ngram",
					"min_gram": 3,
					"max_gram": 5
				}
			}
		}
	},
	"mappings": {
		"properties": {
			"accoundId": {
				"type": "text"
			},
			"email": {
				"type": "text",
				"analyzer": "ngram_filter_analyzer"
			},
			"firstname": {
				"type": "text",
				"analyzer": "edge_ngram_filter_analyzer"
			},
			"surname": {
				"type": "text",
				"analyzer": "edge_ngram_filter_analyzer"
			}
		}
	}
}
```

- We use `ngram_filter_analyzer` for email address.
- We should use `edge_ngram_filter_analyzer` for firstname and surname cause :
  - normally when finding someone's name he would type from the name' first character
  - edge_ngram would reduce the tokens -> reduce the storage need 
  Also, we should support the lower case names. E.g: when I type `bob` I still get `Bob`.
- The `max_ngram_diff` and `min_gram` should be configurable. Either via `elasticsearch.xml` or hard code if we don't need to change them much.

### Query
```json
{
	"query": {
		"bool": {
			"filter": {
				"match": {
					"accountId": "a accoundId"
				}
			},
			"must": {
				"multi_match": {
					"query": "a contact query",
					"type": "bool_prefix",
					"fields": [
						"email",
						"firstname",
						"surname"
					]
				}
			}
		}
	}
}
```

## Domain contact
Basically, the domain contact could be the same, just need to decide to put them into 1 index or 2 indexes.

## TODO
- LAB for domain contact => 1 or 2 index
- Integrate this into the code.

**Comment by chibenwa** [comment]
> Basically, the domain contact could be the same, just need to decide to put them into 1 index or 2 indexes.

One index IMO, we can have part of the id to identify if its a user contact or a domain contact.

**Comment by quantranhong1999** [comment]
> One index IMO, we can have part of the id to identify if its a user contact or a domain contact.

I doubt that it would be harder to CRUD domain contact if we put them into 1 index.

**Comment by chibenwa** [comment]
> I doubt that it would be harder to CRUD domain contact if we put them into 1 index.

One specific search query and we are good?

**Comment by quantranhong1999** [comment]
> One specific search query and we are good?

2 index and the combined query: https://pastebin.com/561YeHyn. It works.
rf: https://discuss.elastic.co/t/query-multiple-indexes-but-apply-queries-to-specific-index/127858/2

**Comment by chibenwa** [comment]
No it looks good tome!

**Comment by quantranhong1999** [comment]
https://github.com/linagora/tmail-backend/pull/389
